mlir/lib/Conversion/AIRToAIEPass.cpp

```diff
@@ -845,6 +846,153 @@ void allocL1Buffers(ModuleOp m,
+    ChannelGetOp get = getTheOtherChannelOpThroughSymbol(put);
+    ChannelOp channel = getChannelDeclarationThroughSymbol(
+        dyn_cast<air::ChannelInterface>(put.getOperation()));
+    if (!channel)
+      return failure();
+    if (!get)
+      return failure();
+
+    // get memrefs of put/get, find memory hierarchy levels
+    auto srcMemref = put.getSrc().getType().cast<MemRefType>();
+    int src_space = srcMemref.getMemorySpaceAsInt();
+    auto dstMemref = get.getDst().getType().cast<MemRefType>();
+    int dst_space = dstMemref.getMemorySpaceAsInt();
+
+    // create aieTiles/shimTiles based on memory hierarchy levels
+    AIE::TileOp producerTile;
+    AIE::TileOp consumerTile;
+    if (src_space == (int)air::MemorySpace::L1 &&
+        dst_space == (int)air::MemorySpace::L1) {
+
+      AIE::CoreOp producerCore = put->getParentOfType<AIE::CoreOp>();
+      if (!producerCore)
+        return failure();
+      producerTile = producerCore.getTileOp();
+      if (!producerTile)
+        return failure();
+
+      AIE::CoreOp consumerCore = get->getParentOfType<AIE::CoreOp>();
+      if (!consumerCore)
+        return failure();
+      consumerTile = consumerCore.getTileOp();
+      if (!consumerTile)
+        return failure();
+
+    } else {
+      assert(false && "Unsupported data movement");
+    }
+
+    // create objFifo
+    // For now, number of memory elements in OF is hardcoded to 1
+    // (single buffer). FIXME
+    rewriter.setInsertionPoint(channel);
+    auto datatype = AIE::AIEObjectFifoType::get(dstMemref);
+    AIE::ObjectFifoCreateOp objFifo =
+        createObjectFifo(rewriter, datatype, producerTile, consumerTile, 1);
+    auto elementType =
+        objFifo.getType().dyn_cast<AIE::AIEObjectFifoType>().getElementType();
+    auto acqType = AIE::AIEObjectFifoSubviewType::get(elementType);
+
+    // replace put and the associated memref alloc
+    if (auto bco =
+            dyn_cast<bufferization::ToMemrefOp>(put.getSrc().getDefiningOp()))
+      rewriter.setInsertionPoint(bco.getOperand().getDefiningOp());
+    else if (auto a = dyn_cast<memref::AllocaOp>(put.getSrc().getDefiningOp()))
+      rewriter.setInsertionPoint(put.getSrc().getDefiningOp());
+    else
+      rewriter.setInsertionPoint(&put->getBlock()->front());
+    AIE::ObjectFifoAcquireOp producerAcq =
+        rewriter.create<AIE::ObjectFifoAcquireOp>(
+            rewriter.getUnknownLoc(), acqType, AIE::ObjectFifoPort::Produce,
+            objFifo, 1);
+    rewriter.setInsertionPointAfter(producerAcq);
+    AIE::ObjectFifoSubviewAccessOp producerAccess =
+        rewriter.create<AIE::ObjectFifoSubviewAccessOp>(
+            rewriter.getUnknownLoc(), elementType, producerAcq.getSubview(),
+            rewriter.getIntegerAttr(rewriter.getI32Type(), 0));
+
+    // replace uses of alloc with result of acquire
+    if (auto a = dyn_cast<memref::AllocOp>(put.getSrc().getDefiningOp()))
+      a.getMemref().replaceAllUsesWith(producerAccess.getOutput());
```

Comment:
Avoid mixing manual IR changes with `rewriter `based changes. The `rewriter` changes are staged and unwound if the pass return `failure()`, but this `replaceAllUsesWith `is immediate.

Reply:
I didn't know this, thank you for explaining! I replaced the function call with one to the rewriter.